proxy/http/HttpTransact.cc

```diff
@@ -1546,6 +1546,28 @@ HttpTransact::OSDNSLookup(State *s)
     TRANSACT_RETURN(SM_ACTION_SEND_ERROR_CACHE_NOOP, nullptr);
   }
 
+  if (s->redirect_info.redirect_in_process && ats_is_ip_loopback(s->host_db_info.ip())) {
+    switch (s->http_config_param->redirect_loopback_mode) {
+      case HttpConfigParams::REDIRECT_LOOPBACK_MODE_REJECT:
+        TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect to loopback address. Rejecting.");
+        build_error_response(s, HTTP_STATUS_FORBIDDEN, nullptr, "request#syntax_error");
+        SET_VIA_STRING(VIA_DETAIL_TUNNEL, VIA_DETAIL_TUNNEL_NO_FORWARD);
+        TRANSACT_RETURN(SM_ACTION_SEND_ERROR_CACHE_NOOP, nullptr);
+        break;
+      case HttpConfigParams::REDIRECT_LOOPBACK_MODE_FOLLOW:
+        // Continue following.
+        TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect to loopback address. Following");
+        break;
+      case HttpConfigParams::REDIRECT_LOOPBACK_MODE_RETURN:
+        // fall-through
+      default:
+        // Return this 3xx to the client as-is.
+        TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect to loopback address. Returning.");
+        build_response_copy(s, &s->hdr_info.server_response, &s->hdr_info.client_response, s->client_info.http_version);
+        TRANSACT_RETURN(SM_ACTION_INTERNAL_CACHE_NOOP, nullptr);
+    }
+  }
```

Comment:
Note to reviewers:

This follows the lazy consensus from the [discussion](https://lists.apache.org/thread.html/c46e9f0493896744ffbfb6eb0d8eed53bb77003a8abbfa1ac5b63c16@%3Cdev.trafficserver.apache.org%3E) on the dev list.

We were sure we wanted this to be configurable, because there are several plausible alternatives.
We were sure the default should not be to follow the redirect.

When the admin configures the proxy to follow up to *N* redirects, if the *N+1*st redirect is to loopback it is not clear what would be expected. Should we simply return the 3xx as-is, or should we stop and inspect its `Location` header to see whether it is loopback and reject it if so?

Since there seemed to be this little ambiguity, I chose the default behavior to simply return the 3xx response to the client. (One can still configure the 403.)

Reply:
Updated as per discussion.